src/NuGet.Services.Staging.Search/DatabaseSearchService.cs

```diff
@@ -69,16 +96,41 @@ internal IReadOnlyList<PackageMetadata> ApplyQueryParameters(int stageKey, bool
             idsResult = ApplyQueryFilter(idsResult, query);
 
             var distinctIds = idsResult.Select(p => p.Id).Distinct().OrderBy(id => id);
-
             var ids = distinctIds.Skip(skip).Take(take);
+
             var idsList = ids.ToList();
 
             var searchResults = ApplyStageFilter(stageKey);
             searchResults = ApplyIncludePrerelease(searchResults, includePrerelease);
 
-            return searchResults.Where(x => idsList.Contains(x.Id)).OrderBy(x => x.Id).ToImmutableList();
+            return searchResults.Where(x => idsList.Contains(x.Id)).OrderBy(x => x.Id).ToList();
+        }
+
+        internal IReadOnlyList<string> AutocompleteInternal(int stageKey, bool includePrerelease, string q, string id, int skip, int take, out int totalHits)
+        {
+            var result = ApplyStageFilter(stageKey);
+            result = ApplyIncludePrerelease(result, includePrerelease);
+
+            var items = ApplyAutocompleteQueryFilter(result, q, id);
+
+            totalHits = items.Count();
+
+            return items.Skip(skip).Take(take).ToList();
         }
 
+        private IQueryable<string> ApplyAutocompleteQueryFilter(IQueryable<PackageMetadata> packages, string q, string id)
+        {
+            if (string.IsNullOrEmpty(q) && string.IsNullOrEmpty(id) || !string.IsNullOrEmpty(q))
+            {
+                // 
+                packages = packages.Where(package => package.Id.Contains(q));

```

Comment:
I can assume here that q is not null, because GetQuery returns string.Empty when q is not specified

---
In reply to: [65842014](https://github.com/NuGet/NuGet.Staging/pull/58#discussion_r65842014) [](ancestors = 65842014)